### PR TITLE
Sphinx

### PR DESCRIPTION
**Pull request by Dorthu**
Moves documentation to using Sphinx.  This is currently a work in progress, although it will likely be merged before full documentation is available.

Remaining items before mergable:

 - [x] Update README detailing how to work with docs
 - [x] Core Concepts page
 - [x] OAuth page
 - [x] Workable navigation
 - [ ] Include autodocs for most models
 - [ ] sphinx directive for model properties

This is done enough for an initial push - I will finish docs for all models shortly.

### CODE REVIEW

File: linode/login_client.py [hunk 1]
```diff
@@ -219,6 +291,28 @@ def _login_uri(self, path):
         return "{}{}".format(self.base_url, path)
 
     def generate_login_url(self, scopes=None, redirect_uri=None):
+        """
+        Generates a url to send users so that they may authenticate to this
```
Comment (yoshizzle) [inline linode/login_client.py]: the description, and function name, calls it a "url" but the params call it a "uri". Inconsistent and should be fixed if possible.

File: linode/objects/filtering.py [hunk 2]
```diff
@@ -1,12 +1,125 @@
...
+
+    :returns: A filter that matches both a and b
+    :rtype: Filter
+    """
     return a.__and__(b)
 
+
+def order_by(field, desc=False):
+    """
+    Allows order of results.  You may only ever order a collection's results
```
Comment (yoshizzle) [inline linode/objects/filtering.py]: Allows ordering of results.

File: linode/paginated_list.py [hunk 3]
```diff
@@ -4,6 +4,33 @@
 
 
 class PaginatedList(object):
+    """
+    The PaginatedList encapsulates the API V4's pagination in an easy to
```
Comment (yoshizzle) [inline linode/paginated_list.py]: in an easily consumable way.

File: linode/paginated_list.py [hunk 4]
```diff
@@ -4,6 +4,33 @@
...
+
+    PaginatedLists should  never be constructed manually, and instead should
+    be created by requesting a collection of resources from the :any:`LinodeClient`.
+    For example::
+
+       linodes = client.linode.get_instances() # returns a PaginatedList of Linodes
+
+    Once you have a PaginatedList of resources, it doesn't matter how many
+    resources the API will return - you can iterate over all of them without
+    worrying about pagination.::
```
Comment (yoshizzle) [inline linode/paginated_list.py]: without having to worry about pagination.

File: linode/paginated_list.py [hunk 5]
```diff
@@ -147,7 +198,8 @@ def make_paginated_list(json, client, cls, parent_id=None, page_url=None,
             filters=None):
         """
         Returns a PaginatedList populated with the first page of data provided,
-        and the ability to load more.
+        and the ability to load more.  This should not be called outside of
```
Comment (yoshizzle) [inline linode/paginated_list.py]: to load additional pages.

File: docs/linode/linode_client.rst [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+   client.linode.get_instances()
+
+See :any:`LinodeClient` for more information on the naming of these groups,
+although generally they are named the same as the first word of the group.
+
+LinodeGroup
+^^^^^^^^^^^
+
+Includes methods for managing and creating Linodes, as well as accessing and
+working with associated services.
```
[the remark this answers is not in the document]
Comment (Dorthu) [reply docs/linode/linode_client.rst]: 😨

File: linode/login_client.py (hunk 1 above)
Comment (Dorthu) [reply linode/login_client.py]: The spec calls it a "redirect uri," hence the naming here, but this function returns a URL (always pointing to https://login.linode.com)..so I'm going to keep it even if it's a little weird.

File: linode/linode_client.py [hunk 7]
```diff
@@ -601,12 +940,34 @@ def create_domain(self, domain, master=True, **kwargs):
...
     def create_volume(self, label, region=None, linode=None, size=20, **kwargs):
         """
-        Creates a new Block Storage Volume, either in the given region, or attached
-        to the given linode.
+        Creates a new Block Storage Volume, either in the given region or
+        attached to the given linode.
+
+        :param label: The label for the new volume.
+        :type label: str
+        :param region: The region to create this volume in.  Not required if
```
Comment (yoshizzle) [inline linode/linode_client.py]: Region

File: linode/linode_client.py [hunk 8]
```diff
@@ -601,12 +940,34 @@ def create_domain(self, domain, master=True, **kwargs):
...
         return self._get_and_filter(Volume, *filters)
 
     def create_volume(self, label, region=None, linode=None, size=20, **kwargs):
         """
-        Creates a new Block Storage Volume, either in the given region, or attached
-        to the given linode.
+        Creates a new Block Storage Volume, either in the given region or
+        attached to the given linode.
+
+        :param label: The label for the new volume.
```
Comment (yoshizzle) [inline linode/linode_client.py]: Volume

File: linode/linode_client.py [hunk 9]
```diff
@@ -601,12 +940,34 @@ def create_domain(self, domain, master=True, **kwargs):
...
+        :returns: A list of Volumes the acting user can access.
+        :rtype: PaginatedList of Volume
+        """
         return self._get_and_filter(Volume, *filters)
 
     def create_volume(self, label, region=None, linode=None, size=20, **kwargs):
         """
-        Creates a new Block Storage Volume, either in the given region, or attached
-        to the given linode.
+        Creates a new Block Storage Volume, either in the given region or
```
Comment (yoshizzle) [inline linode/linode_client.py]: Region

Linode

File: linode/linode_client.py [hunk 10]
```diff
@@ -601,12 +940,34 @@ def create_domain(self, domain, master=True, **kwargs):
...
+
+        :param label: The label for the new volume.
+        :type label: str
+        :param region: The region to create this volume in.  Not required if
+                       `linode` is provided.
+        :type region: Region or str
+        :param linode: The Linode to attach this volume to.  If not given, the
+                       new volume will not be attached to anything.
+        :type linode: Linode or int
+        :param size: The size, in GB, of the new volume.  Defaults to 20.
```
Comment (yoshizzle) [inline linode/linode_client.py]: Volume

File: linode/login_client.py [hunk 11]
```diff
@@ -245,6 +369,20 @@ def finish_oauth(self, code):
...
+        user.  If you simply remove the session your application has for the
+        user without expiring their token, the user is not _really_ logged out.
+
+        :param token: The OAuth token you wish to expire
+        :type token: str
+
+        :returns: If the expiration attempt succeeded.
+        :rtype: bool
+
+        :raises ApiError: If the expiration attempt fails.
```
Comment (yoshizzle) [inline linode/login_client.py]: failed.

File: linode/linode_client.py [hunk 12]
```diff
@@ -423,9 +662,31 @@ def _user_agent(self):
 
     def load(self, target_type, target_id, target_parent_id=None):
         """
-        Constructs and immediately loads the object, circumventing the lazy-loading
-        scheme by immediately making an api request.  Does not load related
-        objects.  Raises an ApiError if the object cannot be loaded.
+        Constructs and immediately loads the object, circumventing the
+        lazy-loading scheme by immediately making an api request.  Does not
```
Comment (yoshizzle) [inline linode/linode_client.py]: API

File: docs/guides/getting_started.rst [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+.. _github: https://github.com/Linode/python-linode-api
+
+.. code-block:: shell
+
+   git clone git@github.com:Linode/python-linode-api
+   cd python-linode-api
+   python setup.py install
+
+.. note::
+   This library uses the "linode" namespace.  This could conflict with other
```
[the remark this answers is not in the document]
Comment (Dorthu) [reply docs/guides/getting_started.rst]: nope, this renders fine

File: linode/linode_client.py [hunk 14]
```diff
@@ -49,25 +97,131 @@ def get_stackscripts(self, *filters, **kwargs):
...
+        :type ltype: str or LinodeType
         :param region: The Region in which we are creating the Linode
-        :param image: The image to deploy to this Linode
-        :param authorized_keys: The ssh public keys to install on the linode's
-                                /root/.ssh/authorized_keys file
-        :param **kwargs: Any other fields to pass to the api
-
-        :returns: A new Linode object
+        :type region: str or Region
+        :param image: The image to deploy to this Linode.  If this is provided
```
Comment (yoshizzle) [inline linode/linode_client.py]: Image

File: linode/linode_client.py [hunk 15]
```diff
@@ -541,11 +818,35 @@ def get_account(self):
...
+        :returns: A list of available Images.
+        :rtype: PaginatedList of Image
+        """
         return self._get_and_filter(Image, *filters)
 
     def create_image(self, disk, label=None, description=None):
         """
         Creates a new Image from a disk you own.
+
+        :param disk: The disk to imagize.
```
Comment (yoshizzle) [inline linode/linode_client.py]: Disk

File: linode/linode_client.py [hunk 16]
```diff
@@ -515,11 +776,23 @@ def delete(self, *args, **kwargs):
 
     # ungrouped list functions
     def get_regions(self, *filters):
+        """
+        Returns the available regions for Linode products.
```
Comment (yoshizzle) [inline linode/linode_client.py]: Regions

File: linode/linode_client.py [hunk 17]
```diff
@@ -515,11 +776,23 @@ def delete(self, *args, **kwargs):
...
+
+        :param filters: Any number of filters to apply to the query.
+
+        :returns: A list of available regions.
```
Comment (yoshizzle) [inline linode/linode_client.py]: Regions

File: linode/linode_client.py [hunk 18]
```diff
@@ -566,12 +867,37 @@ def create_image(self, disk, label=None, description=None):
...
+        :param filters: Any number of filters to apply to this query.
+
+        :returns: A list of NodeBalancers the acting user can access.
+        :rtype: PaginatedList of NodeBalancers
+        """
         return self._get_and_filter(NodeBalancer, *filters)
 
     def create_nodebalancer(self, region, **kwargs):
+        """
+        Creates a new NodeBalancer in the given region.
```
Comment (yoshizzle) [inline linode/linode_client.py]: Region

File: linode/objects/filtering.py [hunk 19]
```diff
@@ -1,12 +1,125 @@
...
+For example, calling :any:`get_instances` returns a list of :any:`Linode`
+objects, so we can use properties of :any:`Linode` to filter the results::
+
+   # returns all Linodes in the "prod" group
+   client.linode.get_instances(Linode.group == "prod")
+
+You can use any boolean comparisons when filtering collections::
+
+   # returns all Linodes _not_ in us-east-1a
+   client.linode.get_instances(Linode.region != "us-east-1a")
```
Comment (yoshizzle) [inline linode/objects/filtering.py]: This one and all the ones below should probably say us-east at this point?

Or are you waiting until next week to be sure?

**Pull request merged**